client/client-lib/src/query.rs

```diff
@@ -23,30 +22,54 @@ impl<R> QueryStrategy<R> for TrustAllPeers {
     }
 }
 
-/// Returns first epoch with a valid sig, otherwise wait till `required` agree
-pub struct ValidHistory {
-    epoch_pk: PublicKey,
-    current: CurrentConsensus<SignedEpochOutcome>,
+/// Returns first response with a valid sig
+#[allow(clippy::type_complexity)]
+pub struct VerifySigned<R> {
+    pk: PublicKey,
+    verifier: Box<dyn Fn(&PublicKey, &R) -> bool + Send + Sync>,
+    allow_consensus_fallback: bool,
+    current: CurrentConsensus<R>,
 }
 
-impl ValidHistory {
-    pub fn new(epoch_pk: PublicKey, required: usize) -> Self {
+impl<R> VerifySigned<R> {
+    /// Strategy for returning first response with a valid sig
+    ///
+    /// * `pk`: Public key used to verify the response
+    /// * `required`: How many responses until a failure or success is returned
+    /// * `allow_consensus_fallback`: Returns a success if cannot verify but `required` agree
+    /// * `verifier`: Function that verifies the data with the public key
+    pub fn new(
+        pk: PublicKey,
+        required: usize,
+        allow_consensus_fallback: bool,
+        verifier: impl Fn(&PublicKey, &R) -> bool + Send + Sync + 'static,
+    ) -> Self {
```

Comment:
nit: `pk` could just be captured by the `verifier` closure, one less arg.

Reply:
Excellent point, this would allow for other kinds of verifiers as well.